lightning-net-tokio/src/lib.rs

```diff
@@ -143,135 +205,135 @@ impl Connection {
-		});
-		tokio::spawn(TcpStream::connect(&addr).select(connect_timeout)
-			.and_then(move |stream| {
-				Connection::setup_outbound(peer_manager, event_notify, their_node_id, stream.0);
-				future::ok(())
-			}).or_else(|_| {
-				//TODO: return errors somehow
-				future::ok(())
-			}));
+	pub async fn connect_outbound<CMH: ChannelMessageHandler + 'static>(peer_manager: Arc<peer_handler::PeerManager<SocketDescriptor, Arc<CMH>>>, event_notify: mpsc::Sender<()>, their_node_id: PublicKey, addr: SocketAddr) {
+		let connect_timeout = time::delay_for(Duration::from_secs(10));
+		let connect_fut = TcpStream::connect(&addr);
+		tokio::select! {
+			_ = connect_timeout => { },
+			res = connect_fut => {
+				if let Ok(stream) = res {
+					Connection::setup_outbound(peer_manager, event_notify, their_node_id, stream).await;
+				}
+			},
+		};
 	}
 }
 
-pub struct SocketDescriptor<CMH: ChannelMessageHandler + 'static> {
+const SOCK_WAKER_VTABLE: task::RawWakerVTable =
+	task::RawWakerVTable::new(clone_socket_waker, wake_socket_waker, wake_socket_waker_by_ref, drop_socket_waker);
+
+fn clone_socket_waker(orig_ptr: *const ()) -> task::RawWaker {
+	descriptor_to_waker(orig_ptr as *const SocketDescriptor)
+}
+fn wake_socket_waker(orig_ptr: *const ()) {
+	wake_socket_waker_by_ref(orig_ptr);
+	drop_socket_waker(orig_ptr);
+}
+fn wake_socket_waker_by_ref(orig_ptr: *const ()) {
+	let descriptor = orig_ptr as *const SocketDescriptor;
+	// An error should be fine. Most likely we got two send_datas in a row, both of which failed to
+	// fully write, but we only need to provide a write_event() once. Otherwise, the sending thread
+	// may have already gone away due to a socket close, in which case there's nothing to wake up
+	// anyway.
+	let mut us = unsafe { (*descriptor).conn.lock() }.unwrap();
+	let _ = us.write_avail.try_send(());
+}
+fn drop_socket_waker(orig_ptr: *const ()) {
+	let _orig_box = unsafe { Box::from_raw(orig_ptr as *mut SocketDescriptor) };
+	// _orig_box is now dropped
+}
+fn descriptor_to_waker(descriptor: *const SocketDescriptor) -> task::RawWaker {
+	let new_box = Box::leak(Box::new(unsafe { (*descriptor).clone() }));
+	let new_ptr = new_box as *const SocketDescriptor;
+	task::RawWaker::new(new_ptr as *const (), &SOCK_WAKER_VTABLE)
+}
+
+pub struct SocketDescriptor {
 	conn: Arc<Mutex<Connection>>,
 	id: u64,
-	peer_manager: Arc<peer_handler::PeerManager<SocketDescriptor<CMH>, Arc<CMH>>>,
 }
-impl<CMH: ChannelMessageHandler> SocketDescriptor<CMH> {
-	fn new(conn: Arc<Mutex<Connection>>, peer_manager: Arc<peer_handler::PeerManager<SocketDescriptor<CMH>, Arc<CMH>>>) -> Self {
+impl SocketDescriptor {
+	fn new(conn: Arc<Mutex<Connection>>) -> Self {
 		let id = conn.lock().unwrap().id;
-		Self { conn, id, peer_manager }
+		Self { conn, id }
 	}
 }
-impl<CMH: ChannelMessageHandler> peer_handler::SocketDescriptor for SocketDescriptor<CMH> {
+impl peer_handler::SocketDescriptor for SocketDescriptor {
```

Comment:
Could you provide a high-level description of this implementation?

Reply:
Just wanted to ping this comment. Trying to get a grasp of how `send_data` works, specifically wrt how `Waker` and `RawWaker` fit in. Feel free to point to a reference.

Reply:
I totally missed this, gotta love the GitHub UI. The only way I know how to surface old comments again is to leave a response as a part of a review, then GH will show that response duplicatively but at the end of the discussion.

In general, Rust futures (which you can mentally model as green threads) work by calling a poll function, which either makes progress or it doesn't. To know when to poll, you provide a waker which receives a call when the runtime (in this case Tokio) thinks polling makes sense now (ie because we have more room in the socket write buffer, as returned by select() etc). Here, we use that to wake the read/process "thread" and then it calls write_buffer_space_avail() which will call send_data() and we'll end up back at poll().

Reply:
It looks like the `ArcWake` trait is just a utility to do essentially what we're doing, but provided by the `futures` crate, which we don't depend on here.

Reply:
Sadly, the way Futures were stabilized upstream is pretty bare-bones. The *only* way to implement your own future is with all the unsafe code and raw function pointers you see here, or via the async keyword on a function.